Proposed fix for routing within controller scope

yoongkang opened the pull request:
See #20312.

This needs some review!

There are two issues I tried to address. I will try my best to describe them.
## Issue 1: Default path helper names

First is the intended behavior for the naming of url helpers for routes defined within a controller block. According to the [docs](http://api.rubyonrails.org/classes/ActionDispatch/Routing.html):

> Note: when using controller, the route is simply named after the method you call on the block parameter rather than map.

``` ruby
# In routes.rb
controller :blog do
  get 'blog/show'     => :list
  get 'blog/delete'   => :delete
  get 'blog/edit/:id' => :edit
end

# provides named routes for show, delete, and edit
link_to @article.title, show_path(id: @article.id)
```

First of all, the docs make little sense here. On one hand it says the route is named after the controller method (BlogController#list in this case) but on the other the default path helper is `show_path`. The path helper also takes an ID parameter, but that isn't reflected in the routes. I am inclined to think that it's intended to be `list_path` as implied by the documentation.

However, this is also problematic, as the path helpers are not namespaced according to the resource name. My belief is this is harmful especially given that we usually have the same method names in different controllers. In this commit, I have changed the default name to (controller_name)_(method_name). For example in this case it will become `blog_list_path`.

If a user wants something that is different from this default, they can pass in an :as parameter.

This particular change caused two other tests to break, which I have also amended.
## Issue 2: The :id thing

In the method `add_route`, the `action` variable (which is the path of the route) is stripped away if an :id parameter is present in the path. So when we pass this action (which is now nil) to the `name_for_action` method we don't get a named url. 

This commit makes a change so that the add_route method uses the `:action` parameter in the `options` hash (as well as the controller from the scope) and uses it to create the action name.

teknull commented:
+1

aruprakshit commented:
+1

sgrif commented:
r? @pixeltrix I don't feel strongly enough about this to merge or close. Do you have an opinion on this change?

pixeltrix commented:
@yoongkang I'm :-1: to using the controller name for the url helper name as that will break a lot of apps for no good reason - in the example you give at the top the url helper generated is actually `blog_show_path` and not `show_path` because it uses the full path to generate the name and not the action name and this is why you had to change the tests. However the `:id` issue could be fixed without breaking existing applications as they don't generate names currently so add defaults would be overridden by existing `:as` options.

I think it'd be good to take the words after the `:id` and use them as prefixes along with singularising the path before the `:id` to get names that match as closely as possible to the names generated by `resources`, e.g:

``` ruby
controller :posts do
  get 'posts'          => :list
  get 'posts/:id'      => :show
  get 'posts/:id/edit' => :edit
end
```

would generate:

```
   Prefix Verb URI Pattern               Controller#Action
    posts GET  /posts(.:format)          posts#list
     post GET  /posts/:id(.:format)      posts#show
edit_post GET  /posts/:id/edit(.:format) posts#edit
```

@yoongkang is this something you could look at doing?

yoongkang committed:
Fix automatic url helper generation

There is an issue that causes automatic url helper generation
to fail when an :id is included in the route.

yoongkang commented:
I've updated the pull request to first fix the `:id` without using the controller name.

I'll make a separate commit to make it match the `resources` behaviour, as that seems to be a different issue.

pixeltrix commented:
@yoongkang sorry, but your change is still a breaking change because you're changing how the name is generated from the path to the action will affect current apps. Please try to make the change without changing any existing tests.

maclover7 commented:
Please rebaes.
